dcore/diagnostics.py

```diff
@@ -18,3 +19,43 @@ def register(self, *fields):
     @staticmethod
     def l2(f):
         return sqrt(assemble(dot(f, f)*dx))
+
+
+class DiagnosticField(object):
+
+    __metaclass__ = ABCMeta
+
+    @abstractproperty
+    def name(self):
+        """The name of this diagnostic field"""
+        pass
+
+    @abstractmethod
+    def compute(self, state):
+        """ Compute the diagnostic field from the current state"""
+        pass
+
+    def __call__(self, state):
+        return self.compute(state)
+
+
+class CourantNumber(DiagnosticField):
+    name = "CourantNumber"
+
+    def area(self, mesh):
+        if hasattr(self, "_area"):
+            return self._area
+        V = FunctionSpace(mesh, "DG", 0)
+        self._area = assemble(TestFunction(V)*dx)
+        return self._area
+
+    def field(self, mesh):
+        if hasattr(self, "_field"):
+            return self._field
+        self._field = Function(FunctionSpace(mesh, "DG", 0), name=self.name)
+        return self._field
```

Comment:
ok, let's construct self._area on the first call, and the UFL expression 
self.exp = TestFunction(V)*dx
and then always do
self._area.assemble(self.exp)
to update